pkg/chunked/filesystem_linux_test.go

```diff
@@ -0,0 +1,360 @@
+	assert.Equal(t, st.Dev, st2.Dev)
+	assert.Equal(t, st.Ino, st2.Ino)
+
+	err = newFile.Close()
+	assert.NoError(t, err)
+}
+
+func TestOpenOrCreateDirUnderRoot(t *testing.T) {
+	rootDir := createTempDir(t)
+	dirName := "dir"
+
+	defer func() {
+		os.RemoveAll(rootDir)
+	}()
+
+	rootFile, err := os.Open(rootDir)
+	assert.NoError(t, err)
+	defer rootFile.Close()
+
+	rootFd := int(rootFile.Fd())
+
+	// try to create a directory outside the root
+	dir, err := openOrCreateDirUnderRoot(rootFd, "../../"+dirName, 0)
+	assert.NoError(t, err)
+	err = dir.Close()
+	assert.NoError(t, err)
+
+	dir, err = openFileUnderRoot(rootFd, dirName, syscall.O_DIRECTORY, 0)
+	assert.NoError(t, err)
+	err = dir.Close()
+	assert.NoError(t, err)
+}
+
+func TestCopyFileContent(t *testing.T) {
+	rootDir := createTempDir(t)
+	defer func() {
+		os.RemoveAll(rootDir)
+	}()
+
+	rootFile, err := os.Open(rootDir)
+	assert.NoError(t, err)
+	defer rootFile.Close()
+
+	rootFd := int(rootFile.Fd())
+
+	file := createTempFile(t, rootDir, "an-existing-file")
+	defer file.Close()
+
+	size, err := file.Write([]byte("Hello, World!"))
+	assert.NoError(t, err)
+
+	_, err = file.Seek(0, io.SeekStart)
+	assert.NoError(t, err)
+
+	st := syscall.Stat_t{}
+	err = syscall.Fstat(int(file.Fd()), &st)
+	assert.NoError(t, err)
+
+	metadata := fileMetadata{
+		FileMetadata: internal.FileMetadata{
+			Name: "new-file",
+			Type: internal.TypeDir,
+		},
+	}
+
+	newFile, newSize, err := copyFileContent(int(file.Fd()), &metadata, rootFd, 0o755, false)
+	assert.NoError(t, err)
+
+	assert.Equal(t, size, int(newSize))
+
+	st2 := syscall.Stat_t{}
+	err = syscall.Fstat(int(newFile.Fd()), &st2)
+	assert.NoError(t, err)
+
+	err = newFile.Close()
+	assert.NoError(t, err)
+
+	// the file was copied without hard links, the inodes must be different
+	assert.Equal(t, st.Dev, st2.Dev)
+	assert.NotEqual(t, st.Ino, st2.Ino)
+
+	metadataCopyHardLinks := fileMetadata{
+		FileMetadata: internal.FileMetadata{
+			Name: "new-file2",
+			Type: internal.TypeDir,
+		},
+	}
+
+	newFile, newSize, err = copyFileContent(int(file.Fd()), &metadataCopyHardLinks, rootFd, 0o755, true)
+	assert.NoError(t, err)
+	assert.Nil(t, newFile)
+
+	// validate it was created as an inode
+	newFile, err = openFileUnderRoot(rootFd, metadataCopyHardLinks.FileMetadata.Name, syscall.O_RDONLY, 0)
+	assert.NoError(t, err)
+
+	assert.Equal(t, size, int(newSize))
+
+	st2 = syscall.Stat_t{}
+	err = syscall.Fstat(int(newFile.Fd()), &st2)
+	assert.NoError(t, err)
+
+	err = newFile.Close()
+	assert.NoError(t, err)
+
+	// the file was copied with hard links, the inodes must be equal
+	assert.Equal(t, st.Dev, st2.Dev)
+	assert.Equal(t, st.Ino, st2.Ino)
+}
+
+func createTempDir(t *testing.T) string {
+	tmpDir, err := os.MkdirTemp("", "testDir")
+	assert.NoError(t, err)
+	return tmpDir
+}
+
+func createTempFile(t *testing.T, dir, name string) *os.File {
+	tmpFile, err := os.CreateTemp(dir, name)
+	assert.NoError(t, err)
```

Comment:
(Test case setup should typically use `require.NoError` when there’s no point in proceeding. But that can always be fixed later, if the test ever actually fail.)